File: tp1_grafos/novo_campus/final.py
import sys
from collections import defaultdict


class MinHeap:
    def __init__(self):
        self.elements = []

    def pop(self):
        if len(self.elements) == 0:
            return [None, None]

        if len(self.elements) == 1:
            return self.elements.pop()

        root_value = self.elements[0]
        self.elements[0] = self.elements.pop()
        self.bubble_down(0)

        return root_value

    def bubble_down(self, index):
        left_child_index = 2 * index + 1
        right_child_index = 2 * index + 2
        min_index = index

        if left_child_index < len(self.elements) and self.elements[left_child_index][0] < self.elements[min_index][0]:
            min_index = left_child_index

        if (
            right_child_index < len(self.elements)
            and self.elements[right_child_index][0] < self.elements[min_index][0]
        ):
            min_index = right_child_index

        if min_index != index:
            self.elements[index], self.elements[min_index] = (
                self.elements[min_index],
                self.elements[index],
            )
            self.bubble_down(min_index)

    def push(self, priority, value):
        self.elements.append((priority, value))
        self.bubble_up(len(self.elements) - 1)

    def bubble_up(self, index):
        if index == 0:
            return

        parent_index = (index - 1) // 2
        parent_priority = self.elements[parent_index][0]
        current_priority = self.elements[index][0]

        if parent_priority > current_priority:
            self.elements[parent_index], self.elements[index] = (
                self.elements[index],
                self.elements[parent_index],
            )
            self.bubble_up(parent_index)

    def peek(self):
        if len(self.elements) == 0:
            return None

        return self.elements[0][1]


class Graph:
    def __init__(self, num_v):
        self.num_v = num_v
        self.adj = defaultdict(lambda: defaultdict(int))
        self.key = {}
        self.out_cut = {}
        for vertex in range(num_v):
            self.out_cut[vertex + 1] = True
        self.queue = MinHeap()

    def add_edge(self, u, v, w_u_v):
        if u not in self.adj.keys():
            self.key[u] = float("inf")
        if v not in self.adj.keys():
            self.key[v] = float("inf")
        self.adj[u][v] = w_u_v
        self.adj[v][u] = w_u_v

    def run_mst(self, start_vertex):
        self.key[start_vertex] = 0
        self.queue.push(0, start_vertex)
        self.out_cut[start_vertex] = False
        while self.queue.elements:
            _, current_vertex = self.queue.pop()
            self.out_cut[current_vertex] = False
            neighbors_list = self.adj[current_vertex].keys()
            for neighbor in neighbors_list:
                neigh_cost = self.adj[current_vertex][neighbor]
                if self.out_cut[neighbor] and self.key[neighbor] > neigh_cost:
                    self.key[neighbor] = self.adj[current_vertex][neighbor]
                    self.queue.push(neigh_cost, neighbor)

        total_cost = sum(self.key.values())
        if True in self.out_cut.values():
            total_cost = "impossivel"
        return total_cost


if __name__ == "__main__":
    data = sys.stdin.read().splitlines()
    iterator = iter(data)

    while True:
        try:
            info = next(iterator)
        except:
            break

        num_v, num_e = info.split(" ")
        graph = Graph(int(num_v))
        for edge_idx in range(int(num_e)):
            field_info = next(iterator)
            vertex_u, vertex_v, cost = field_info.split(" ")
            graph.add_edge(int(vertex_u), int(vertex_v), int(cost))

        result = graph.run_mst(1)
        sys.stdout.write(str(result))

File: tp1_grafos/final_mundial_de_2008/final.cpp
// O Código foi baseado em uma implementação feita em Python.
// Contudo, como a execução no Python estava demorando mais do que aceito pela plataforma, foi feita uma tradução para C++.
// A lógica do algoritmo permanece a mesma, a única mudança foi a linguagem implementada, aparentemente o Python não conseguiu comportar o limite de 2s.
// import sys
// import heapq
// from collections import defaultdict


// class Graph:
//     def __init__(self, num_v):
//         self.num_v = num_v
//         self.adj = defaultdict(lambda: defaultdict(int))
//         self.distances = {}
//         self.unvisited = set()
//         self.pi = {}
//         self.queue = []

//     def add_edge(self, u, v, w_u_v):
//         if u not in self.adj:
//             self.distances[u] = float("inf")
//             self.unvisited.add(u)
//         if v not in self.adj:
//             self.distances[v] = float("inf")
//             self.unvisited.add(v)

//         self.adj[u][v] = w_u_v

//     def reset_dijkstra(self):
//         self.pi.clear()
//         self.queue.clear()

//     def run_dijkstra(self, origin, destiny, t_allowed):
//         unvisited = self.unvisited.copy()
//         distances = self.distances.copy()
//         if origin == destiny:
//             return 0
//         elif t_allowed == 0:
//             if destiny in self.adj.get(origin, []):
//                 direct_connection = self.adj[origin][destiny]
//             else:
//                 direct_connection = -1

//             return direct_connection

//         distances[origin] = 0
//         heapq.heappush(self.queue, (0, origin))
//         distances[origin] = 0

//         while self.queue:
//             min_element = heapq.heappop(self.queue)
//             if min_element[1] not in unvisited:
//                 continue
//             current_vertex_coast, current_vertex = min_element
//             unvisited.remove(current_vertex)

//             if current_vertex == destiny:
//                 break

//             neighbors_list = self.adj[current_vertex].keys()

//             for neighbor in neighbors_list:
//                 if neighbor in unvisited and (neighbor <= t_allowed or neighbor == destiny):
//                     neigh_cost = self.adj[current_vertex][neighbor]
//                     neigh_total_coast = current_vertex_coast + neigh_cost
//                     if neigh_total_coast < distances[neighbor]:
//                         distances[neighbor] = neigh_total_coast
//                         self.pi[neighbor] = current_vertex
//                         heapq.heappush(self.queue, (neigh_total_coast, neighbor))

//         if distances[destiny] == float("inf"):
//             return -1
//         else:
//             return distances[destiny]


// if __name__ == "__main__":
//     data = sys.stdin.read().splitlines()
//     iterator = iter(data)
//     instance_counter = 1

//     while True:
//         try:
//             info = next(iterator)
//             sys.stdout.write(f"Instancia {instance_counter}\n")
//         except:
//             break

//         num_v, num_e = info.split(" ")
//         graph = Graph(int(num_v))
//         for edge_idx in range(int(num_e)):
//             field_info = next(iterator)
//             vertex_u, vertex_v, cost = list(map(int, field_info.split(" ")))
//             if vertex_v not in graph.adj.get(vertex_u, []):
//                 graph.add_edge(vertex_u, vertex_v, cost)
//             else:
//                 current_coast = graph.adj.get(vertex_u, {}).get(vertex_v, float("inf"))
//                 if cost < current_coast:
//                     graph.add_edge(vertex_u, vertex_v, cost)

//         eval_cases = int(next(iterator).strip())
//         for eval_c in range(eval_cases):
//             origin, destiny, t = next(iterator).split(" ")
//             result = graph.run_dijkstra(int(origin), int(destiny), int(t))
//             graph.reset_dijkstra()
//             sys.stdout.write(f"{result}\n")

//         sys.stdout.write("\n")
//         instance_counter += 1



#include <bits/stdc++.h>
using namespace std;

static constexpr long long INF = 1000000000000000LL; // 1e15

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    int instancia = 1;

    while ( (cin >> n >> m) ) {
        vector<vector<pair<int,int>>> adj(n+1);
        {
            vector<unordered_map<int,int>> tmp(n+1);
            for (int e = 0; e < m; ++e) {
                int u, v, w;
                cin >> u >> v >> w;
                auto it = tmp[u].find(v);
                if (it == tmp[u].end() || w < it->second) tmp[u][v] = w;
            }
            for (int u = 1; u <= n; ++u) {
                adj[u].reserve(tmp[u].size());
                for (auto &kv : tmp[u]) {
                    adj[u].push_back({kv.first, kv.second});
                }
            }
        }

        int q;
        cin >> q;

        cout << "Instancia " << instancia++ << "\n";

        for (int qi = 0; qi < q; ++qi) {
            int o, d, t;
            cin >> o >> d >> t;

            if (o == d) {
                cout << 0 << "\n";
                continue;
            }
            if (t == 0) {
                long long best = INF;
                for (auto [v, w] : adj[o]) if (v == d) best = min(best, (long long)w);
                cout << (best < INF ? best : -1) << "\n";
                continue;
            }

            vector<long long> dist(n+1, INF);
            vector<char> visited(n+1, 0);
            priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<pair<long long,int>>> pq;

            auto can_enter = [&](int v) -> bool {
                return (v == d) || (v <= t);
            };

            dist[o] = 0;
            pq.push({0LL, o});

            while (!pq.empty()) {
                auto [du, u] = pq.top(); pq.pop();
                if (visited[u]) continue;
                visited[u] = 1;

                if (u == d) break;

                for (auto [v, w] : adj[u]) {
                    if (!can_enter(v)) continue;
                    long long nd = du + (long long)w;
                    if (nd < dist[v]) {
                        dist[v] = nd;
                        pq.push({nd, v});
                    }
                }
            }

            cout << (dist[d] < INF ? dist[d] : -1) << "\n";
        }
        cout << "\n";
    }
    return 0;
}
